evm-contracts/src/v0.6/examples/VRFD20.sol

```diff
@@ -0,0 +1,118 @@
+pragma solidity 0.6.6;
+
+import "../VRFConsumerBase.sol";
+import "../Owned.sol";
+
+/**
+ * @notice A Chainlink VRF consumer which uses randomness to mimic the rolling
+ * of a 20 sided die
+ */
+contract VRFD20 is VRFConsumerBase, Owned {
+
+    bytes32 private s_keyHash;
+    uint256 private s_fee;
+
+    mapping(address => bytes32) private s_rollers;
+    mapping(bytes32 => uint256) private s_results;
+
```

Comment:
@se3000 how does this look?

The requestId needs to be a key for the fulfil function, and the address of the roller need to be a key so that they can query their result/d&d character.

Reply:
I think that on the lookup it would be more efficient to read the address directly instead of pulling the request ID and then using that to look up the address. So it would be:
```
    mapping(address => bytes32) private s_rollers;
    mapping(address => uint256) private s_results;
```
Then when you want to roll, you just make sure that the there is no roll or result present for the address.

Reply:
You could possibly consolidate the above two mappings into a single struct, but I could see the argument that this is simpler too.

Reply:
The problem with your example is the fulfillment is difficult. The address is not involved in the fulfillment process, so setting `s_results` when the dice has landed will involve searching through all address keys in `s_rollers` to find the correct `requestId` value, then update `s_results`.

Reply:
... unless we only allow 1 roll to be in progress at any given time and lock the contract up during it.

Reply:
I've kept this the same as fulfilment will be inefficient otherwise.

Reply:
Good point about retrieving the address on fulfillment, I missed that. Still, the final read, `s_houses[s_results[s_rollers[player]]]`, requires 3 reads from storage and the s_houses read is extra large because it's a string; that's a lot of gas for a read.

How about:
```
mapping(bytes32 => address) private s_requests;
mapping(address => uint256) private s_results;
```
When the roll happens, mark the address's `s_results` with some special number like 42, which is outside the range of dice rolls and just indicates a roll is in progress. Then, create a helper function which just has an array of strings in memory and looks up the string based on index.